### ISSUE TITLE
Modify-Filter insanely broken?

### ISSUE THREAD

**Issue by jesk78** [fluent/fluent-bit#1077]
## Bug Report

**Describe the bug**

Modify Filter deletes records totally unexpectedly and sticks to a single Filter instance for the whole fluent-bit runtime. 

_**Have a look at the first GitHub comment for a more detailed description.**_

**To Reproduce**

```
[INPUT]
    Name tail
    Refresh_Interval 3
    Buffer_Chunk_Size 1000M
    Buffer_Max_Size 1000M
    Path /var/log/foo.json
    DB /var/log/foo.db
    Tag foo
    Parser json

[FILTER]
    Name modify
    Match foo*
    Condition Key_value_equals foo_status 64
    Add foo_status_enc nothing_else_gets_through

[OUTPUT]
    Name es
    Match foo*
    Host 127.0.0.1
    Port 9200
    Logstash_Format true
    Logstash_Prefix foo
    Logstash_DateFormat %Y.%m.%d
    Type _doc
```

**Expected behavior**

Modify filter should add a key:"foo_status_enc" with value:"nothing_else_gets_through" when the record has a key:"foo_status" with value:"64". It should not filter out (drop) records not having that key/value (foo_status/64).

**Environment**

CentOS Linux release 7.2.1511 (Core)
td-agent-bit.x86_64    1.0.2-1
and
td-agent-bit.x86_64    1.0.4-1

**Comment by jesk78** [comment]
**I just built/tried the current master branch (1.1.0) with the same result:**

**Setup:**
Multiple modify filter instances, each having key_value_equals condition on same key but on different value and all are adding the same key/field but with different values.

**Behavior / bugs:**
**(1)** Records which do *not* match any of those modify-filters will get lost
**(2)** The modify-filter which will match the first record after fluent-bit has been started will be the only one which will match ever again. This in combination with the behavior of **(1)** leads to a massive data  deletion of everything but those records, which do have the same key/value pair as the first record which was matched by one of those filter instances. Only those will make it to the OUTPUT, everything else -> /dev/null.

Something seems to be insanely broken.


**Example configuration:**

```
[SERVICE]
    Flush             5
    Parsers_File      parsers.conf
    Log_Level         info
    Log_File          /var/log/fluentbit.log

[INPUT]
    Name              tail
    Refresh_Interval  3
    Buffer_Chunk_Size 1000M
    Buffer_Max_Size   1000M
    Path              /var/log/netflow.json
    DB                /var/log/netflow.db
    Tag               netflow
    Parser            json

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 0
    Add       fwd_status_enc unknown

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 64
    Add       fwd_status_enc forwarded_unknown

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 65
    Add       fwd_status_enc forwarded_frag

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 66
    Add       fwd_status_enc forwarded_non_frag

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 128
    Add       fwd_status_enc dropped_unknown

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 129
    Add       fwd_status_enc dropped_acl_deny

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 130
    Add       fwd_status_enc dropped_acl_drop

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 131
    Add       fwd_status_enc dropped_unrouteable

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 132
    Add       fwd_status_enc dropped_adjacency

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 133
    Add       fwd_status_enc dropped_frag_needed_df

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 134
    Add       fwd_status_enc dropped_bad_hdr_chksum

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 135
    Add       fwd_status_enc dropped_bad_total_len

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 136
    Add       fwd_status_enc dropped_bad_hdr_len

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 137
    Add       fwd_status_enc dropped_bad_ttl

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 138
    Add       fwd_status_enc dropped_policer

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 139
    Add       fwd_status_enc dropped_wred

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 140
    Add       fwd_status_enc dropped_rpf

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 141
    Add       fwd_status_enc dropped_for_us

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 142
    Add       fwd_status_enc dropped_bad_out_intf

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 143
    Add       fwd_status_enc dropped_hardware

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 192
    Add       fwd_status_enc consumed_unknown

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 193
    Add       fwd_status_enc consumed_punt_adj

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 194
    Add       fwd_status_enc consumed_incompl_adj

[FILTER]
    Name      modify
    Match     netflow
    Condition Key_value_equals fwd_status 195
    Add       fwd_status_enc consumed_for_us

[OUTPUT]
    Name                es
    Match               netflow
    Host                127.0.0.1
    Port                9200
    Logstash_Format     true
    Logstash_Prefix     netflow
    Logstash_DateFormat %Y.%m.%d
    Type                _doc
```

**Comment by edsiper** [comment]
I did a local test and it worked, would you please share your test file with some lines?

**Comment by jesk78** [comment]
Please try with that sample data:
[netflow_sample.zip](https://github.com/fluent/fluent-bit/files/2837962/netflow_sample.zip)

**Comment by jesk78** [comment]
Here some simple test configuration:

`[SERVICE]
    Parsers_File      parsers.conf
    Log_Level         trace

[INPUT]
    Name              tail
    Path              test.json
    Tag               test
    Parser            json

[FILTER]
    Name      modify
    Match     test
    Condition Key_value_equals mickey mouse
    Add       do like

[FILTER]
    Name      modify
    Match     test
    Condition Key_value_equals donald duck
    Add       do likeaswell

[OUTPUT]
    Name  stdout
    Match *
`
test.conf:
`{"mickey":"mouse"}
{"donald":"duck"}
{"mickey":"goofie"}
{"mickey":"mouse"}
`

**Comment by jesk78** [comment]
Is really noone impacted by this? Confusing.

**Comment by jesk78** [comment]
I'm using Lua now, seems modify filter can be marked as dead and broken.

**Comment by grigorov** [comment]
@jesk78 can you share your Lua script?

**Comment by jesk78** [comment]
function addfscode(tag, timestamp, record)
	local map = {
		  ["0"] = "unknown",
		 ["64"] = "forwarded_unknown",
		 ["65"] = "forwarded_frag",
		 ["66"] = "forwarded_non_frag",
		["128"] = "dropped_unknown",
		["129"] = "dropped_acl_deny",
		["130"] = "dropped_acl_drop",
		["131"] = "dropped_unrouteable",
		["132"] = "dropped_adj",
		["133"] = "dropped_frag_df",
		["134"] = "dropped_bad_hdr_chksum",
		["135"] = "dropped_bad_total_len",
		["136"] = "dropped_bad_hdr_len",
		["137"] = "dropped_bad_ttl",
		["138"] = "dropped_policer",
		["139"] = "dropped_wred",
		["140"] = "dropped_rpf",
		["141"] = "dropped_for_us",
		["142"] = "dropped_bad_out_intf",
		["143"] = "dropped_hardware",
		["192"] = "consumed_unknown",
		["193"] = "consumed_punt_adj",
		["194"] = "consumed_incompl_adj",
		["195"] = "consumed_for_us",
	}
	if record["fwd_status"] ~= nil then
		local fscode = record["fwd_status"]
		if map[fscode] ~= nil then
			record["fwd_status_enc"] = map[fscode]
			return 1, timestamp, record
		else
			return 0, 0, 0
		end
	else
		return 0, 0, 0
	end
end

function addifalias(tag, timestamp, record)
	local map = {}
	map["192.0.2.1"] = {
		[37] = "foo",
		[38] = "bar",
		[39] = "omg",
		[40] = "wtf",
		[41] = "bbq",
	}
	map["192.0.2.2"] = {
		[21] = "abc",
		[22] = "def",
		[24] = "ghi",
		[17] = "jkl",
		[18] = "mno",
		[19] = "pqr",
		[59] = "stu",
		[60] = "vwx",
		[61] = "yz!",
	}
	local ip = record["peer_ip_src"]
	local idxin  = record["iface_in"]
	local idxout = record["iface_out"]
	if map[ip][idxin] ~= nil then
		record["ifalias_in"] = map[ip][idxin]
	end
	if map[ip][idxout] ~= nil then
		record["ifalias_out"] = map[ip][idxout]
	end
	return 1, timestamp, record
end

**Comment by vezir** [comment]
Is there any improvement in this issue?

**Comment by nokute78** [comment]
I sent a patch to fix this issue. (#1416)

The output will be like this.
```
[0] test: [1562233919.205549498, {"mickey"=>"mouse", "do"=>"like"}]
[1] test: [1562233919.205562366, {"donald"=>"duck", "do"=>"likeaswell"}]
[2] test: [1562233919.205564829, {"mickey"=>"goofie"}]
[3] test: [1562233919.205566708, {"mickey"=>"mouse", "do"=>"like"}]
```

**Comment by edsiper** [comment]
fixed by #1416
